trieste/utils/pareto.py

```diff
@@ -148,3 +149,44 @@ def hypervolume_indicator(self, reference: TensorType) -> TensorType:
         hypervolume = tf.reduce_sum(tf.reduce_prod(upper - lower, 0))
 
         return tf.reduce_prod(reference[None] - min_pfront) - hypervolume
+
+    def hypercell_bounds(
+        self, anti_reference: TensorType, reference: TensorType
+    ) -> tuple[TensorType, TensorType]:
+        """
+        Get the partitioned hypercells lower and upper bounds
+        :param anti_reference: a worst point to use with shape [D].
+            Defines the lower bound of the hypercell
+        :param reference: a reference point to use, with shape [D].
+            Defines the upper bound of the hypervolume.
+            Should be equal or bigger than the anti-ideal point of the Pareto set.
+            For comparing results across runs, the same reference point must be used.
+        :return: lower, upper bounds or the partitioned cell
+        :raise ValueError (or `tf.errors.InvalidArgumentError`): If ``reference`` has an invalid
+            shape.
+        """
+        tf.debugging.assert_greater_equal(reference, self.front)
+        tf.debugging.assert_greater_equal(self.front, anti_reference)
+
+        tf.debugging.assert_shapes(
+            [
+                (self.bounds.lower_idx, ["N", "D"]),
+                (self.bounds.upper_idx, ["N", "D"]),
+                (self.front, ["M", "D"]),
+                (reference, ["D"]),
+                (anti_reference, ["D"]),
+            ]
+        )
+
+        pseudo_pfront = tf.concat((anti_reference[None], self.front, reference[None]), axis=0)
```

Comment:
@joelberkeley one thing not very sure here is, do we need some defensive programming to cast dtype here?

Reply:
it's a good point. I'm not sure what's best. The alternative would be raise an exception if the dtypes didn't match. While that's a little strict, it's what I'd expect tensorflow to do so the consistency might help. You could do that with `tf.debugging.assert_same_float_dtype`

Reply:
Used `tf.debugging.assert_type` instead.
Seems `tf.debugging.assert_same_float_dtype` do not support dtype assert when Tensor's shape>1, weird.